File: app/src/main/java/com/coinninja/coinkeeper/cn/wallet/tx/TransactionFundingManager.java
package com.coinninja.coinkeeper.cn.wallet.tx;

import com.coinninja.bindings.DerivationPath;
import com.coinninja.bindings.TransactionData;
import com.coinninja.bindings.UnspentTransactionOutput;
import com.coinninja.coinkeeper.cn.account.AccountManager;
import com.coinninja.coinkeeper.cn.wallet.HDWallet;
import com.coinninja.coinkeeper.di.interfaces.TransactionDust;
import com.coinninja.coinkeeper.model.db.Address;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary;
import com.coinninja.coinkeeper.model.db.TargetStat;
import com.coinninja.coinkeeper.model.helpers.InviteTransactionSummaryHelper;
import com.coinninja.coinkeeper.model.helpers.TargetStatHelper;
import com.coinninja.coinkeeper.service.client.model.TransactionFee;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class TransactionFundingManager {
    private final HDWallet hdWallet;
    private final long valueOfDust;
    private final AccountManager accountManager;
    private final TargetStatHelper targetStatHelper;
    private final InviteTransactionSummaryHelper inviteTransactionSummaryHelper;
    private List<UnspentTransactionOutput> unspentTransactionOutputs;

    @Inject
    TransactionFundingManager(HDWallet hdWallet, @TransactionDust long valueOfDust, AccountManager accountManager,
                              TargetStatHelper targetStatHelper, InviteTransactionSummaryHelper inviteTransactionSummaryHelper) {

        this.hdWallet = hdWallet;
        this.valueOfDust = valueOfDust;
        this.accountManager = accountManager;
        this.targetStatHelper = targetStatHelper;
        this.inviteTransactionSummaryHelper = inviteTransactionSummaryHelper;
    }

    public TransactionData buildFundedTransactionData(TransactionFee transactionFee) {
        TransactionData transactionData = createTransactionData();
        if (transactionFee == null) return transactionData;

        long fees = 0;
        long changeAmount = 0;
        long walletValue = calculateSpendableBalance();
        long amountSending = 0;

        if (walletValue > 0) {
            fees = buildInputsWithFees(transactionFee, walletValue);
            amountSending = walletValue - fees;
            changeAmount = calculateChange(walletValue);
        }

        if (amountSending > 0) {
            transactionData.setAmount(amountSending);
            transactionData.setFeeAmount(fees);
            transactionData.setChangeAmount(changeAmount);
            transactionData.setUtxos(getFundingInputs());
        }

        return transactionData;
    }

    public TransactionData buildFundedTransactionData(TransactionFee transactionFee, long amountToSend) {
        TransactionData transactionData = createTransactionData();
        if (transactionFee == null) return transactionData;

        long fees = 0;
        long changeAmount = 0;
        long walletValue = calculateSpendableBalance();

        if (walletValue > amountToSend) {
            fees = buildInputsWithFees(transactionFee, amountToSend);
            changeAmount = calculateChange(amountToSend + fees);
        }

        if (walletValue >= amountToSend + fees) {
            transactionData.setAmount(amountToSend);
            transactionData.setFeeAmount(fees);
            transactionData.setChangeAmount(changeAmount);
            transactionData.setUtxos(getFundingInputs());
        }

        return transactionData;
    }

    public TransactionData buildFundedTransactionDataForDropBit(long amountToSend, long explicitFee) {
        TransactionData transactionData = createTransactionData();
        long changeAmount = 0;
        long walletValue = calculateValueOfWallet();
        long totalTransactionCost = amountToSend + explicitFee;

        if (walletValue > 0) {
            changeAmount = calculateChange(totalTransactionCost);
        }

        if (walletValue >= totalTransactionCost) {
            transactionData.setAmount(amountToSend);
            transactionData.setFeeAmount(explicitFee);
            transactionData.setChangeAmount(changeAmount);
            transactionData.setUtxos(getFundingInputs());
        }

        return transactionData;
    }

    protected long calculateSpendableBalance() {
        return calculateValueOfWallet() + calculateValueOfPendingDropbits();
    }

    protected long buildInputsWithFees(TransactionFee transactionFee, long amountToSend) {
        long transactionFeePerInput = calculateFeePerInput(transactionFee);
        List<UnspentTransactionOutput> requiredInputs = new ArrayList<>();
        int inputCount = 0;
        int outputCount = 1;
        long inputsValue = 0L;
        long currentFee = 0L;
        long totalSendingValue;

        for (UnspentTransactionOutput unspentTransactionOutput : getFundingInputs()) {
            totalSendingValue = amountToSend + currentFee;

            if (totalSendingValue > inputsValue) {
                requiredInputs.add(unspentTransactionOutput);
                inputCount += 1;
                inputsValue += unspentTransactionOutput.getAmount();
                currentFee = transactionFeePerInput * (inputCount + outputCount);
                totalSendingValue = amountToSend + currentFee;

                long changeValue = inputsValue - totalSendingValue;
                if (changeValue > 0 && changeValue < transactionFeePerInput + getValueOfDust()) {
                    currentFee += changeValue;
                    break;
                } else if (changeValue > 0) {
                    currentFee += transactionFeePerInput;
                    break;
                }
            }
        }
        setRequiredInputs(requiredInputs);
        return currentFee;
    }

    protected UnspentTransactionOutput[] getFundingInputs() {
        UnspentTransactionOutput[] outputs = new UnspentTransactionOutput[unspentTransactionOutputs.size()];
        return unspentTransactionOutputs.toArray(outputs);
    }

    protected long calculateFeePerInput(TransactionFee transactionFee) {
        return hdWallet.getFeeInSatoshis(transactionFee, 1);
    }


    protected long calculateValueOfWallet() {
        long value = 0L;
        for (TargetStat stat : targetStatHelper.getSpendableTargets()) {
            value += stat.getValue();
            unspentTransactionOutputs.add(stat.toUnspentTranasactionOutput());
        }
        return value;
    }

    protected long calculateValueOfPendingDropbits() {
        long value = 0L;
        for (InviteTransactionSummary dropbit : inviteTransactionSummaryHelper.getUnfulfilledSentInvites()) {
            value -= dropbit.getValueFeesSatoshis();
            value -= dropbit.getValueSatoshis();
        }
        return value;
    }

    @NotNull
    protected TransactionData createTransactionData() {
        unspentTransactionOutputs = new ArrayList<>();
        return new TransactionData(new UnspentTransactionOutput[0], 0, 0, 0,
                nextChangePath(), null);
    }

    protected DerivationPath nextChangePath() {
        Address address = new Address();
        address.setIndex(accountManager.getNextChangeIndex());
        address.setChangeIndex(HDWallet.INTERNAL);
        return address.getDerivationPath();
    }

    protected long calculateChange(long totalTransactionCost) {
        long value = 0L;
        for (UnspentTransactionOutput unspentTransactionOutput : unspentTransactionOutputs) {
            value += unspentTransactionOutput.getAmount();
        }

        return value - totalTransactionCost;
    }

    protected long getValueOfDust() {
        return valueOfDust;
    }

    protected void setRequiredInputs(List<UnspentTransactionOutput> requiredInputs) {
        unspentTransactionOutputs.clear();
        unspentTransactionOutputs.addAll(requiredInputs);
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/cn/wallet/tx/TransactionFundingManager.kt
package com.coinninja.coinkeeper.cn.wallet.tx

import app.coinninja.cn.libbitcoin.enum.ReplaceableOption
import app.coinninja.cn.libbitcoin.model.TransactionData
import app.coinninja.cn.libbitcoin.model.UnspentTransactionOutput
import app.dropbit.annotations.Mockable
import javax.inject.Inject

@Mockable
class TransactionFundingManager @Inject internal constructor(
        val fundingModel: FundingModel) {

    data class BuiltInput(val fees: Long = 0, val utxos: Array<UnspentTransactionOutput> = emptyArray()) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as BuiltInput

            if (fees != other.fees) return false
            if (!utxos.contentEquals(other.utxos)) return false

            return true
        }

        override fun hashCode(): Int {
            var result = fees.hashCode()
            result = 31 * result + utxos.contentHashCode()
            return result
        }
    }

    fun buildFundedTransactionData(payToAddress: String?, feeRate: Double): TransactionData {
        val transactionData = createTransactionData()
        val walletValue = fundingModel.spendableAmount
        val builtInput: BuiltInput = if (walletValue > 0) {
            buildInputsWithFees(payToAddress, feeRate, walletValue)
        } else BuiltInput()

        val amountSending = walletValue - builtInput.fees

        if (amountSending > 0) {
            transactionData.apply {
                amount = amountSending
                feeAmount = builtInput.fees
                utxos = builtInput.utxos
                paymentAddress = payToAddress
            }
        }

        return transactionData
    }

    fun buildFundedTransactionData(payToAddress: String?, feeRate: Double, amountToSend: Long, rbf: Boolean? = null): TransactionData {
        val transactionData = createTransactionData()
        val walletValue = fundingModel.spendableAmount

        val builtInput: BuiltInput = if (walletValue > 0) {
            buildInputsWithFees(payToAddress, feeRate, amountToSend)
        } else BuiltInput()

        if (walletValue >= amountToSend + builtInput.fees) {
            transactionData.apply {
                amount = amountToSend
                feeAmount = builtInput.fees
                utxos = builtInput.utxos
                changeAmount = calculateChange(amountToSend, builtInput.fees, builtInput.utxos)
                utxos = builtInput.utxos
                rbf?.let {
                    replaceableOption = if (rbf) ReplaceableOption.RBF_ALLOWED else ReplaceableOption.MUST_NOT_BE_RBF
                }
            }
        }

        return transactionData
    }

    fun buildFundedTransactionDataForDropBit(payToAddress: String?, amountToSend: Long, explicitFee: Long): TransactionData {
        val transactionData = createTransactionData()
        val walletValue = fundingModel.spendableAmount
        val totalTransactionCost = amountToSend + explicitFee

        val input = buildInputsWithFees(payToAddress, 0.0, amountToSend, explicitFee = explicitFee)

        if (walletValue >= totalTransactionCost) {
            transactionData.apply {
                amount = amountToSend
                feeAmount = explicitFee
                changeAmount = calculateChange(totalTransactionCost, utxos = input.utxos)
                utxos = input.utxos
                replaceableOption = ReplaceableOption.MUST_BE_RBF
            }
        }

        return transactionData
    }

    fun isTransactionFundableWithFee(address: String?, amountToSend: Long, feeRate: Double): Boolean {
        return buildFundedTransactionData(address, feeRate, amountToSend).amount > 0
    }


    internal fun buildInputsWithFees(payToAddress: String?, feeRate: Double, amountToSend: Long, explicitFee: Long? = null): BuiltInput {
        val unspentTransactionOutputs: MutableList<UnspentTransactionOutput> = mutableListOf()
        var inputsValue: Long = 0
        var txBytes = FundingModel.baseTransactionBytes + fundingModel.outPutSizeInBytesForAddress(payToAddress)
        var currentFee: Long = explicitFee ?: 0
        var totalSendingValue: Long

        fundingModel.unspentTransactionOutputs.forEach funding@{ unspentTransactionOutput ->
            totalSendingValue = amountToSend + currentFee

            if (totalSendingValue > inputsValue) {
                unspentTransactionOutputs.add(unspentTransactionOutput)
                inputsValue += unspentTransactionOutput.amount
                txBytes += fundingModel.inputSizeInBytes
                currentFee = explicitFee ?: fundingModel.calculateFeeForBytes(txBytes, feeRate)
                totalSendingValue = amountToSend + currentFee

                val changeValue = inputsValue - totalSendingValue
                if (changeValue > 0 && changeValue < costOFPotentialChangeAndDust(feeRate)) {
                    currentFee += changeValue
                    return@funding
                } else if (changeValue > 0) {
                    txBytes += fundingModel.changeSizeInBytes
                    currentFee = explicitFee ?: fundingModel.calculateFeeForBytes(txBytes, feeRate)
                    return@funding
                }
            }
        }
        return BuiltInput(currentFee, unspentTransactionOutputs.toTypedArray())
    }

    private fun costOFPotentialChangeAndDust(feeRate: Double) =
            fundingModel.calculateFeeForBytes(fundingModel.changeSizeInBytes, feeRate) + fundingModel.transactionDustValue


    protected fun createTransactionData(): TransactionData {
        return TransactionData(emptyArray(), 0, 0, 0,
                fundingModel.nextChangePath, "")
    }

    protected fun calculateChange(totalTransactionCost: Long, fees: Long = 0, utxos: Array<UnspentTransactionOutput> = emptyArray()): Long {
        var value = 0L
        utxos.forEach { output ->
            value += output.amount
        }
        return value - totalTransactionCost - fees
    }
}
